lib/u2/ToggleListElementsView.es6.js

```diff
@@ -0,0 +1,234 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  name: 'ToggleListElementsView',
+  package: 'org.chromium.apis.web',
+  extends: 'foam.u2.Element',
+
+  documentation: `A view for toggling a list of arbitrary elements. Similar to
+      foam.u2.view.EditColumns view, but with a more consistent treatment of
+      list elements and the slots that bind them.
+
+      Elements in the selectable and selected lists are identified by their name
+      and use their label for rendering in UI. As such, list elements must each
+      have both a unique name and a label.`,
+
+  requires: [
+    'foam.u2.md.CheckBox'
+  ],
+  imports: ['error'],
+  exports: [
+    'selectable',
+    'selected',
+  ],
+
+  css: `
+    ^item {
+      display: flex;
+    }
+
+    ^ .foam-u2-md-CheckBox {
+      flex-shrink: 0;
+    }
+
+    ^ .foam-u2-md-CheckBox-label {
+      position: static;
+    }
+  `,
+
+  classes: [
+    {
+      name: 'ElementSlot',
+      extends: 'foam.core.Slot',
+
+      documentation: `A custom foam.core.Slot implementation; each manages one
+          element in the selected collection.`,
+
+      imports: [
+        'selectable',
+        'selected',
+      ],
+
+      properties: [
+        {
+          class: 'Int',
+          name: 'selectableIdx',
+        },
+      ],
+
+      methods: [
+        function get() {
+          const elem = this.selectable[this.selectableIdx];
+          return this.selected.some(c => c.name === elem.name);
+        },
+        function set(value) {
+          const prevValue = this.get();
+          if (value) {
+            this.addElement();
+          } else {
+            this.removeElement();
+          }
+          // Data binding depnds on an event being published on set. This is
+          // automatically taken care of by property slots (with the
+          // propertyChange event), but must be published manually for this
+          // custom foam.core.Slot implementation.
+          if (prevValue !== value) {
+            this.pub('elementChange', this.selectableIdx, this);
+          }
+        },
+        function getElement() {
+          return this.selectable[this.selectableIdx];
+        },
+        function addElement() {
+          const elem = this.selectable[this.selectableIdx];
+          const able = this.selectable;
+          const ed = this.selected;
+          let iAble = 0;
+          let iEd = 0;
+          while (iAble <= this.selectableIdx) {
```

Comment:
An easier to follow implementation could be to store the selected state of each element in the element, and then compute the selected array by filtering selectable on it for each update. That also prevents having duplicates added to selected, which this doesn't appear to guard against.

Reply:
This implementation was intended to be non-intrusive to the elements of `selected` and `selectable`. I'll have to think about the duplicate guard issue, though.

Reply:
Added multiple-addElement() guard and more algorithm documentation. @arobins PTAL.